sparta/sparta/resources/Scoreboard.hpp

```diff
@@ -0,0 +1,306 @@
+#pragma once
+
+#include <vector>
+#include <cinttypes>
+#include <string>
+#include <map>
+#include <list>
+
+#include "sparta/simulation/ParameterSet.hpp"
+#include "sparta/simulation/Unit.hpp"
+#include "sparta/simulation/TreeNode.hpp"
+#include "sparta/events/PayloadEvent.hpp"
+
+namespace sparta
+{
+    class ScoreboardView;
+
+    /**
+     * \class Scoreboard
+     *
+     * The Scoreboard of the model simply keeps track of the readiness
+     * of phyiscal registers in the OOO core.  There are two parts to the SB:
+     *
+     * # The Scoreboard or "master" for each register file type (GPU,
+     *   FPR, Vector, etc).  Typically a Rename block is responsible
+     *   for setting/clearing the SB readiness.
+     *
+     * # The ScoreboardView is created by a Scheduling/Execution block
+     *   and is used to determine if an instruction is ready for
+     *   execution (all operands ready)
+     *
+     */
+    class Scoreboard : public sparta::Unit
```

Comment:
I'm thinking (for now) that this derive from `sparta::Resource`.  It's possible that a modeler might have a special logger/stat set for this and would prefer to pass them in.

Reply:
Question: how a modeler pass logger/stat in? Does modeler defined them in the derive class inheriting Scoreboard?

Reply:
One way to do it is for the main scoreboard class to be dynamically constructed during the parent resource's constructor.  Since this is allowed in Sparta, the main constructor of the Scoreboard can either:
1. Look for it's parent's StatSet/Loggers 
2. If the parent _does not_ have a statset/logger, the SB creates its own and uses those

Reply:
The Scoreboard views should be instantiated in the constructor of the using class, which will have statset/logger already -- in theory.

Reply:
@knute-sifive Gotcha. But I haven't figured out what's the benefit to use parent's StatSet/Loggers. Wouldn't it be clearer that having messages from Scoreboard instead of parent resource? l'll defer it until we run into that need.

Reply:
Couple of notes on the subject:
1. With the Scoreboard implementation as it is, the loggers are forced to `info`, `debug`, etc from `sparta::Unit`.  If I create a Scoreboard in my unit, the SB will be a sub-unit and logging enablement will _always include_ scoreboard output.  That would get verbose REALLY fast
2. If the SB has an API to support a separate logger, I can put the scoreboard on a "logger" scoreboard and separate out the messages.